hpy/devel/__init__.py

```diff
@@ -293,10 +327,28 @@ def _finalize_hpy_ext(self, ext):
         ext.name = HPyExtensionName(ext.name)
         ext.hpy_abi = self.distribution.hpy_abi
         ext.include_dirs += self.hpydevel.get_extra_include_dirs()
-        ext.sources += self.hpydevel.get_extra_sources()
+        # If we may use static libs (default), then add all available libs (for
+        # the given ABI) to the extra objects. The libs will then just be added
+        # in the linking phase but nothing will be compiled in addition.
+        use_static_libs = not self.distribution.hpy_no_static_libs
+        if use_static_libs:
+            static_libs = self.hpydevel.get_static_libs(ext.hpy_abi)
+        else:
+            static_libs = None
+
+        if static_libs:
+            ext.extra_objects += static_libs
+        else:
+            # If we should not use (pre-compiled) static libs or if they are
+            # not available, we just add the sources of the helpers to the
+            # extension. They are then compiler with the extension.
+            ext.sources += self.hpydevel.get_extra_sources()
```

Comment:
so, if I understand correctly the logic is the following:
1. by default we use the `.a` file
2. if you specify `--hpy-no-static-libs`, we use `*.c` files
3. if you try to use the `.a` file but you don't find it, you fallback to (2)

I think that (1) and (2) are reasonable, but what is the use case for (3)?
In which case do we expect it to be useful?

If we don't have a concrete use case and the logic is there "just in case", I vote for removing it and emit a clean error instead (something like "Cannot find libhpy.a"): basically I'm worried about the following scenario:
1. at some point the future we make a mistake when building `libhpy*.a.` (e.g. by putting it in the wrong directory or something like that) 
2. we don't notice because things still work as usual (but using *.c instead of the static lib)
3. at some point someone will try to compile a C++ lib again and they will be very confused

Reply:
To elaborate more: basically, instead of "let's search from random *.a files in the given lib", I'd change the logic to "let's find hpy/devel/lib/libhpy-extra.a" (or libhpy-extra-cpython.a, accordingly).

Reply:
Sorry that I never answered your comments. I was side-tracked and just forgot about it.

> so, if I understand correctly the logic is the following:
> - by default we use the .a file
> - if you specify --hpy-no-static-libs, we use *.c files
> - if you try to use the .a file but you don't find it, you fallback to (2)

Yes, right now, it tries to find `*.a` (or `*.lib` on Windows) files for the given HPy ABI.
That can be disabled using `--hpy-no-static-libs`.
If the libs are not available but we want use static libs (your point (3) ), we fallback and include/compile sources with the extension as before.
Reason: I tried to be defensive and didn't want to make things worse as before. It is IMO even necessary because we only ship libraries for our well-known ABIs. If someone has another ABI than `universal` or `cpython`, then we cannot ship the static libs.

But I also agree, it might be confusing. Anyway, I think I'll go a step back with that PR and will just use this for the tests.
I realized that there is a severe conceptual problem: shipping the static libs means that they are built in the CI. But the whole point of the extra sources was that they are compiled on the *client* side for portability (because of varargs etc).

> To elaborate more: basically, instead of "let's search from random *.a files in the given lib", I'd change the logic to a more explicit "let's find hpy/devel/lib/libhpy-extra.a" (or libhpy-extra-cpython.a, accordingly).

I tried to avoid looking for a specific file like `libhpy-extra.a` because setuptools' support for that is poor. We would need to make assumptions about the naming scheme and I tried to avoid that by just looking into a well-known folder (which is ABI-specific) and take whatever is there.